### PR TITLE
Use dockerized Lambda.

### PR DESCRIPTION
**Pull request by carsten-bo**
*Issue:* #6 

*Description of changes:* 
- Adds Dockerfile for creating a dockerized Lambda image
- Changes Cloudformation Template to use Image URI instead of 

In order to being able to distribute the Lambda in a better way, I created a  Dockerfile that builds the image with the 'AWS Lambda Python Runtime Interface Client'. This allows to provide the URI to the Cloudformation Template and deploy the Lambda without having to create the zip file and add it to S3. 

See the example Docker image that was automatically built with Dockerhub using my fork: https://hub.docker.com/r/carstenbo/amazon-athena-cross-account-catalog

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    --no-cache \
+    openssl \
+    openjdk8-jre
+
+COPY --from=build-image /opt/venv /opt/venv
+ENV PATH="/opt/venv/bin:$PATH"
+
+COPY --from=build-image heracles ./heracles
+
+#(Optional) Add Lambda Runtime Interface Emulator and use a script in the ENTRYPOINT for simpler local runs
```
Comment (dacort) [inline Dockerfile]: Why not just use the official `public.ecr.aws/lambda/python:3.9` image? It includes the emulator by default and should have everything we need to run this code. In addition, using that would probably cut the Dockerfile down to 10-15 lines.

File: entry.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+if [ -z "${AWS_LAMBDA_RUNTIME_API}" ]; then
+    exec /usr/bin/aws-lambda-rie /opt/venv/bin/python -m awslambdaric $1
+else
+    exec /opt/venv/bin/python -m awslambdaric $1
+fi
```
Comment (dacort) [inline entry.sh]: This would be unnecessary if we used the official AWS base image.

**Review by dacort — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (carsten-bo) [reply Dockerfile]: You're right, the complexity of the Dockerfile would be way lower. I tested with this and it also works:

```
FROM public.ecr.aws/lambda/python:3.8

COPY requirements.txt .
RUN python3 -m pip install -r requirements.txt && rm requirements.txt

COPY heracles ./heracles

CMD [ "heracles.lambda.handler" ]
```

But the image is almost triple the size at about 600mb. I like smaller images, but I agree it adds a lot of complexity to the Dockerfile. So, your call :).

By the way, the 3.9 tag is not yet released for the base image.

Comment (carsten-bo) [reply Dockerfile]: @dacort I changed it now to use the official base image for easier maintainability.

**Comment by carsten-bo** [comment]
I changed my master to have only relevant commits, sorry for the mess :sweat_smile:

**Comment by carsten-bo** [comment]
I think it might make more sense to not overwrite the existing CF template (for now) but add a separate one to be able to decide which type of deployment one would prefer. If you want to keep only the image deployment option the Readme, would need to be reworked completely. I now only updated it to show the other option of deployment.

**Comment by dacort** [comment]
Awesome, thanks @carsten-bo! I appreciate all the changes and moving to the official image. I like the smaller size, but for this example I think I prefer the easy maintainability. :)

**Pull request merged**